plugin/trino-iceberg/src/main/java/io/trino/plugin/iceberg/IcebergSplitSource.java

```diff
@@ -13,91 +13,228 @@
+        private final long dynamicFilteringProbeBlockingTimeoutMillis;
+        private final Stopwatch stopwatch;
+
+        public IcebergSplitLoader(
+                SchemaTableName schemaTableName,
+                Set<IcebergColumnHandle> identityPartitionColumns,
+                CloseableIterable<CombinedScanTask> combinedScanIterable,
+                DynamicFilter dynamicFilter,
+                TimeZoneKey timeZoneKey,
+                AsyncQueue<ConnectorSplit> splitQueue,
+                long dynamicFilteringProbeBlockingTimeoutMillis)
+        {
+            this.schemaTableName = requireNonNull(schemaTableName, "schemaTableName is null");
+            this.identityPartitionColumns = requireNonNull(identityPartitionColumns, "identityPartitionColumns is null");
+            this.combinedScanIterable = requireNonNull(combinedScanIterable, "combinedScanIterable is null");
+            this.dynamicFilter = requireNonNull(dynamicFilter, "dynamicFilter is null");
+            this.timeZoneKey = requireNonNull(timeZoneKey, "timeZoneKey is null");
+            this.splitQueue = requireNonNull(splitQueue, "splitQueue is null");
+            this.dynamicFilteringProbeBlockingTimeoutMillis = dynamicFilteringProbeBlockingTimeoutMillis;
+
+            this.fileScanIterator = Streams.stream(combinedScanIterable)
+                    .map(CombinedScanTask::files)
+                    .flatMap(Collection::stream)
+                    .iterator();
+            this.stopwatch = Stopwatch.createStarted();
+        }
+
+        @Override
+        public TaskStatus process()
+        {
+            long probeTimeRemaining = dynamicFilteringProbeBlockingTimeoutMillis - stopwatch.elapsed(MILLISECONDS);
+            if (dynamicFilter.isAwaitable() && probeTimeRemaining > 0) {
+                return TaskStatus.continueOn(asVoid(toListenableFuture(dynamicFilter.isBlocked().orTimeout(probeTimeRemaining, MILLISECONDS))));
+            }
+
+            while (fileScanIterator.hasNext()) {
+                FileScanTask task = fileScanIterator.next();
+                if (!task.deletes().isEmpty()) {
+                    throw new TrinoException(NOT_SUPPORTED, "Iceberg tables with delete files are not supported: " + schemaTableName);
+                }
+                IcebergSplit icebergSplit = toIcebergSplit(task);
+                if (partitionPassesDynamicFilter(icebergSplit.getPartitionKeys())) {
+                    ListenableFuture<Void> queueOfferFuture = splitQueue.offer(toIcebergSplit(task));
+                    if (!queueOfferFuture.isDone()) {
+                        // Queue is full, wait until there is space
+                        return TaskStatus.continueOn(queueOfferFuture);
+                    }
+                }
+            }
+            splitQueue.finish();
+            return TaskStatus.finished();
+        }
+
+        public boolean partitionPassesDynamicFilter(Map<Integer, String> partitionKeys)
```

Comment:
Can we have common code for this and `TableStatisticsMaker#dataFileMatches` ?

Reply:
Can we also look at file level Domain for the columns for which we have dynamic filter and prune splits based on that ?
We don't necessarily have to restrict ourselves to partitioned columns here.